aws/image/misc-settings.sh

```diff
@@ -20,24 +20,105 @@ if [ "$CLOUD_PROVIDER" == "vsphere" ]
 then
 # Add vsphere specific commands to execute on remote
     case $PODVM_DISTRO in
-	centos)
-	    #fallthrough
-	    ;&
-	rhel)
-	    (! dnf list --installed | grep open-vm-tools >/dev/null 2>&1) && \
-		(! dnf -y install open-vm-tools) && \
-		     echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2 && exit 1
-	    ;;
-	ubuntu)
-	    (! dpkg -l | grep open-vm-tools > /dev/null 2>&1) && apt-get update && \
-		(! apt-get -y install open-vm-tools > /dev/null 2>&1) && \
-		     echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2  && exit 1
-	    ;;
-	*)
-	    ;;
+    centos)
+        #fallthrough
+        ;&
+    rhel)
+        (! dnf list --installed | grep open-vm-tools > /dev/null 2>&1) && \
+        (! dnf -y install open-vm-tools) && \
+             echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2 && exit 1
+        ;;
+    ubuntu)
+        (! dpkg -l | grep open-vm-tools > /dev/null 2>&1) && apt-get update && \
+        (! apt-get -y install open-vm-tools > /dev/null 2>&1) && \
+             echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2  && exit 1
+        ;;
+    *)
+        ;;
     esac
-# else if...
-#
+fi
+
+# Setup oneshot systemd service for AWS and Azure to enable NAT rules
+if [ "$CLOUD_PROVIDER" == "azure" ] || [ "$CLOUD_PROVIDER" == "aws" ]
+then
+# Create a oneshot systemd service to execute setup-nat-for-imds.sh
+# during first boot
+cat <<END > /etc/systemd/system/setup-nat-for-imds.service
+[Unit]
+Description=Setup NAT for IMDS
+After=kata-agent.service
+Wants=kata-agent.service
+Type=oneshot
+
+[Service]
+ExecStart=/usr/local/bin/setup-nat-for-imds.sh
+
+[Install]
+WantedBy=multi-user.target
+END
+
+# Add a link to the service in multi-user.target.wants
+ln -s /etc/systemd/system/setup-nat-for-imds.service /etc/systemd/system/multi-user.target.wants/setup-nat-for-imds.service
+
+# Create a script to setup NAT for IMDS
+cat <<END > /usr/local/bin/setup-nat-for-imds.sh
+#!/bin/bash
+# This script sets up NAT for IMDS on Azure/AWS
+# This is required for IMDS to work on Azure/AWS
+# This script is executed as a oneshot systemd service
+# during first boot
+# Define variables
+
+set -euo pipefail
+
+POD_NS="podns"
+IMDS_IP_CIDR="169.254.169.254/32"
+VETH_NET_CIDR="192.168.100.0/30"
+VETH1_IP_CIDR="192.168.100.1/30"
+VETH2_IP_CIDR="192.168.100.2/30"
+VETH_GW="192.168.100.1"
+
```

Comment:
I think exposing these specific IP addresses to workloads may cause conflicts with user's network.

Another possible approach is [Proxy ARP](https://en.wikipedia.org/wiki/Proxy_ARP).

The following network configuration works in my environment, although it looks somewhat complicated.

```
imds_ip="169.254.169.254/32"
pod_ip=$(jq -r '.["pod-network"]["podip"]' /peerpod/daemon.json )
dummy_ip="169.254.99.99/32"

ip link add veth2 type veth peer name veth1
ip address add "$dummy_ip" dev veth1 # Proxy arp does not get enabled when no IP address is assigned
ip link set up dev veth1

sysctl -w net.ipv4.conf.veth1.proxy_arp=1
sysctl -w net.ipv4.conf.veth1.proxy_delay=0

ip link set veth2 netns podns
ip netns exec podns ip link set up dev veth2
ip netns exec podns ip route add "$imds_ip" dev veth2

ip route add "$pod_ip" dev veth1

iptables -t nat -A POSTROUTING -s "$pod_ip" -d "$imds_ip" -j MASQUERADE
```

With this configuration, no additional IP address is visible to workloads as follows.

```
podvm# ip netns exec podns ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
3: vxlan0@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UNKNOWN group default
    link/ether c6:2f:dc:9e:bb:26 brd ff:ff:ff:ff:ff:ff link-netns hostns
    inet 172.20.254.185/32 scope global vxlan0
       valid_lft forever preferred_lft forever
    inet6 fe80::c42f:dcff:fe9e:bb26/64 scope link
       valid_lft forever preferred_lft forever
13: veth2@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 5a:d2:d6:ec:b6:aa brd ff:ff:ff:ff:ff:ff link-netns hostns
    inet6 fe80::58d2:d6ff:feec:b6aa/64 scope link
       valid_lft forever preferred_lft forever
podvm# ip netns exec podns ip r
default via 169.254.1.1 dev vxlan0
169.254.1.1 dev vxlan0 scope link
169.254.169.254 dev veth2 scope link
```

Reply:
I like that! proxy-arp seems well-suited to our specific problem here. 👍

Reply:
I was failing to get this working on rhel, i hit several issues,
1. jq was missing
2. the pod-ip mask caused to "Invalid prefix for given prefix length"
3. proxy_delay configuration didn't exist

Reply:
Looks like the proxy_delay path is `net.ipv4.neigh.veth1.proxy_delay`

Reply:
I updated the script to address some problems.

```
iimds_ip="169.254.169.254"
pod_ip=$(ip netns exec podns ip route get 169.254.169.254 | awk 'NR == 1 { print $7 }')
dummy_ip="169.254.99.99"

ip link add veth2 type veth peer name veth1
ip address add "$dummy_ip/32" dev veth1 # Proxy arp does not get enabled when no IP address is assigned
ip link set up dev veth1

sysctl -w net.ipv4.conf.veth1.proxy_arp=1
sysctl -w net.ipv4.neigh.veth1.proxy_delay=0

ip link set veth2 netns podns
ip netns exec podns ip link set up dev veth2
ip netns exec podns ip route add "$imds_ip/32" dev veth2

ip route add "$pod_ip/32" dev veth1

hwaddr=$(ip netns exec podns ip -br link show veth2 | awk 'NR==1 { print $3 }')
ip neigh replace "$pod_ip" dev veth1 lladdr "$hwaddr"

iptables -t nat -A POSTROUTING -s "$pod_ip/32" -d "$imds_ip/32" -j MASQUERADE
```

Reply:
This is not working with OVN-kubernetes. Says "no route found".
The previous approach of using local IP addresses seems to be deterministic.

Reply:
I was gonna suggest instead of using the `192.xx.` series using `169.254.0.1/31` and `169.254.0.2/31`. Assuming that this won't create conflicts.

Reply:
@surajssd the approach suggested by @yoheiueda is different (uses proxy_arp) and it works fine